### PR TITLE
Add preliminary support for love.js as a build target

### PR DESCRIPTION
**Pull request by jiaaro**

### CODE REVIEW

**Comment by pfirsich** [comment]
Hey! This looks really exciting! I wonder if it makes more sense to call the target `lovejs` instead of web, in case some other system will be supported in the future either alongside love.js or replacing it, but having incompatible configuration.

**Comment by jiaaro** [comment]
works for me

File: makelove/web.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    except URLError as exc:
+        eprint("Could not download löve: {}".format(exc))
+        eprint(
+            "If there is in fact no download on GitHub for this version, specify 'love_binaries' manually."
+        )
+        sys.exit(1)
+    print("Download complete")
+
+
+def render_mustache(tmpl, cx):
```
Comment (pfirsich) [inline makelove/web.py]: That name confuses me. Is that a thing? Searching in both love.js repos for "mustache" gives me nothing. What is it? Should it maybe have a comment?

**Review by pfirsich — CHANGES_REQUESTED** [review]
I'm not really sure if the steps you are following here are correct. I just had a couple minutes look at the love.js repos and I'm not sure where you got your steps from. Can you link to something? Or did you figure it out yourself? In that case we might want to fix the love.js-version so it stays functional?

**Comment by pfirsich** [comment]
I will have another look and test it tomorrow. It's a national holiday, so I will have no excuse not to!

**Comment by pfirsich** [comment]
Would you believe it. I don't have any project to test this on, because my 11.3-projects use LuaJIT-features (either the bit module or FFI). I think you should allow building projects for 11.0, 11.1 and 11.2, because 11.3 should be backwards compatible and maybe only display a warning, if it doesn't match (or ask with a yes/no-prompt, but I'm not sure if I like that).

File: makelove/web.py (hunk 1 above)
Comment (jiaaro) [reply makelove/web.py]: lovejs uses [mustache templates](https://github.com/janl/mustache.js/) I didn't want to introduce a dependency nor implement the whole template system, so I just implemented the variable interpolation, which is all lovejs uses anyway.

File: makelove/web.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        else:
+            download_love(config["love_version"], target)
+
+    src = os.path.join(love_binaries, "love.zip")
+    dst = os.path.join(target_directory, f"{config['name']}-{target}.zip")
+    with open(src, "rb") as lovef, ZipFile(lovef) as love_binary_zip, open(
+        dst, "wb+"
+    ) as outf, ZipFile(outf, mode="w") as app_zip, open(
+        love_file_path, "rb"
+    ) as love_zip:
```
[the remark this answers is not in the document]
Comment (jiaaro) [reply makelove/web.py]: totally agree. but also, ideally, I won't need to add 5 levels of indentation to open 3 files 😂

File: makelove/web.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+from pathlib import Path
+from zipfile import ZipFile
+from urllib.request import urlretrieve, URLError
+
+from .util import eprint, get_default_love_binary_dir, get_download_url
+
+
+def download_love(version, platform):
+    """
+    Note, mac builds are stored as zip files because extracting them
```
[the remark this answers is not in the document]
Comment (jiaaro) [reply makelove/web.py]: yeah, caught me haha

**Comment by jiaaro** [comment]
> I'm not really sure if the steps you are following here are correct. I just had a couple minutes look at the love.js repos and I'm not sure where you got your steps from. Can you link to something? Or did you figure it out yourself? In that case we might want to fix the love.js-version so it stays functional?

The steps in the love.js repo are for people who want to run the tool using nodejs - however lovejs primarily provides a set of JS files which are the love 11.3 repo compiled into web assembly. I've reimplemented the templating lovejs does in javascript in python (see their index.js). Everything else is just copying their compiled files to the right places

File: makelove/web.py (hunk 1 above)
Comment (pfirsich) [reply makelove/web.py]: Could you add this as a comment? Because I see myself forgetting that and wondering the same thing again in the future.

**Comment by pfirsich** [comment]
I tried out some projects. Simple love projects that don't do anything fancy work fine (this worked: https://pfirsich.itch.io/spacetris and it's amazing), but love.js can't do LuaJIT specific stuff, which includes FFI, but also the `bit` module. And obviously you can't add any Lua modules from shared libraries, which shockingly excludes almost every other project I have (damn)!

I wonder if we should include a note about this somewhere and maybe warn/error if the configuration includes shared libraries.
I added a note to the readme, but maybe there needs to be more? 

I attached the few missing changes as a patch to this comment, because I didn't want to PR in your repo (I think that would end up in a total mess). I don't know how to handle this properly. You don't have to commit it yourself (you can though!). I just want to make sure you agree with the changes: [pull_9.txt](https://github.com/pfirsich/makelove/files/5397454/pull_9.txt) (I have to attach as .txt, because GitHub won't let me upload a .diff or .patch :))

File: makelove/web.py (hunk 1 above)
Comment (pfirsich) [reply makelove/web.py]: I did add the comment in the patch I attached to my last comment on this PR

**Comment by pfirsich** [comment]
After a short private conversation it turned out jiaaro is pretty busy right now, so I will just merge and add the final adjustments.

**Pull request merged**